test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Metadata/DefaultModelMetadataTest.cs

```diff
@@ -909,6 +909,325 @@ public void GetMetadataForProperties_CallsProvider()
+            var result = DefaultModelMetadata.TryCalculateHasValidators(new HashSet<DefaultModelMetadata>(), modelMetadata, out var hasValidators);
+
+            // Assert
+            Assert.True(result);
+            Assert.False(hasValidators);
+        }
+
+        [Fact]
+        public void TryCalculateHasValidators_SelfReferencingType_HasValidatorOnNestedProperty()
+        {
+            // Arrange
+            var modelType = typeof(Employee);
+            var modelIdentity = ModelMetadataIdentity.ForType(modelType);
+            var metadataProvider = new Mock<IModelMetadataProvider>();
+            var modelMetadata = CreateModelMetadata(modelIdentity, metadataProvider.Object, false);
+
+            var employeeId = ModelMetadataIdentity.ForProperty(typeof(int), nameof(Employee.Id), modelType);
+            var employeeIdMetadata = CreateModelMetadata(modelIdentity, metadataProvider.Object, false);
+            var employeeUnit = ModelMetadataIdentity.ForProperty(typeof(BusinessUnit), nameof(Employee.Unit), modelType);
+            var employeeUnitMetadata = CreateModelMetadata(employeeUnit, metadataProvider.Object, false);
+            var employeeManager = ModelMetadataIdentity.ForProperty(typeof(Employee), nameof(Employee.Unit), modelType);
+            var employeeManagerMetadata = CreateModelMetadata(employeeManager, metadataProvider.Object, false);
+            var employeeEmployees = ModelMetadataIdentity.ForProperty(typeof(List<Employee>), nameof(Employee.Employees), modelType);
+            var employeeEmployeesMetadata = CreateModelMetadata(employeeEmployees, metadataProvider.Object, false);
+
+            var unitHead = ModelMetadataIdentity.ForProperty(typeof(Employee), nameof(BusinessUnit.Head), modelType);
+            var unitHeadMetadata = CreateModelMetadata(unitHead, metadataProvider.Object, false);
+            var unitId = ModelMetadataIdentity.ForProperty(typeof(int), nameof(BusinessUnit.Id), modelType);
+            var unitIdMetadata = CreateModelMetadata(unitId, metadataProvider.Object, true); // BusinessUnit.Id has validators.
+
+            metadataProvider
+                .Setup(mp => mp.GetMetadataForProperties(modelType))
+                .Returns(new[] { employeeIdMetadata, employeeUnitMetadata, employeeManagerMetadata, employeeEmployeesMetadata, })
+                .Verifiable();
+
+            metadataProvider
+                .Setup(mp => mp.GetMetadataForProperties(typeof(BusinessUnit)))
+                .Returns(new[] { unitHeadMetadata, unitIdMetadata, })
+                .Verifiable();
+
+            // Act
+            var result = DefaultModelMetadata.TryCalculateHasValidators(new HashSet<DefaultModelMetadata>(), modelMetadata, out var hasValidators);
+
+            // Assert
+            Assert.True(result);
+            Assert.True(hasValidators);
+        }
+
+        [Fact]
+        public void TryCalculateHasValidators_SelfReferencingType_HasValidatorOnSelfReferencedProperty()
```

Comment:
We should also test cycles where the cycle length > 1 and also cycles that involve a collection.